packages/astro/src/actions/runtime/virtual/server.ts

```diff
@@ -212,3 +230,80 @@ function unwrapBaseObjectSchema(schema: z.ZodType, unparsedInput: FormData) {
 	}
 	return schema;
 }
+
+/**
+ * Access information about Action requests from middleware.
+ */
+export function getMiddlewareContext(context: APIContext) {
+	const callerInfo = getCallerInfo(context);
+
+	// Prevents action results form being handled on a rewrite.
+	// Also prevents our *own* fallback middleware from running
+	// if the user's middleware has already handled the result.
+	const actionResultAlreadySet = Boolean((context.locals as Locals)._actionPayload);
+
+	let action:
+		| {
+				/** Whether an action was called using an RPC function or by using an HTML form action. */
+				calledFrom: 'rpc' | 'form-action';
+				/** The name of the action. Useful to track the source of an action result during a redirect. */
+				name: string;
+				/** Programatically call the action to get the result. */
+				handler: () => Promise<SafeResult<any, any>>;
+		  }
+		| undefined = undefined;
+
+	if (callerInfo && context.request.method === 'POST' && !actionResultAlreadySet) {
+		const contentType = context.request.headers.get('content-type');
+		const contentLength = context.request.headers.get('Content-Length');
+
+		action = {
+			calledFrom: callerInfo.from,
+			name: callerInfo.name,
+			handler: async () => {
+				const baseAction = await getAction(callerInfo.name);
+				let input: unknown = undefined;
+
+				if (!contentType || contentLength === '0') {
+					input = undefined;
+				} else if (contentType && hasContentType(contentType, formContentTypes)) {
+					input = await context.request.clone().formData();
+				} else if (contentType && hasContentType(contentType, ['application/json'])) {
+					input = await context.request.clone().json();
+				} else {
+					return { data: undefined, error: new ActionError({ code: 'UNSUPPORTED_MEDIA_TYPE' }) };
+				}
+				const handler = baseAction.bind(context);
+				return handler(input);
+			},
+		};
+	}
+
+	/**
+	 * Manually set the action result accessed via `getActionResult()`.
+	 * Calling this function from middleware will disable Astro's own action result handling.
+	 */
+	function setActionResult(actionName: string, actionResult: SerializedActionResult) {
+		(context.locals as Locals)._actionPayload = {
+			actionResult,
+			actionName,
+		};
+	}
+	return {
+		action,
+		setActionResult,
+		serializeActionResult,
+		deserializeActionResult,
+	};
+}
+
+function getCallerInfo(ctx: APIContext) {
+	if (ctx.routePattern === '/_actions/[...path]') {
```

Comment:
I believe is safer to create a const at this point, since we have the same path somewhere else.